### PR TITLE
[ISSUE-#4467] Fix snowflake id problem

### PR DESCRIPTION
**Pull request by horizonzy**
## What is the purpose of the change

Fix-#4467

### CODE REVIEW

File: naming/src/main/java/com/alibaba/nacos/naming/core/ServiceManager.java [hunk 1]
```diff
@@ -811,7 +811,15 @@ public Instance getInstance(String namespaceId, String serviceName, String clust
...
-                instance.setInstanceId(instance.generateInstanceId(currentInstanceIds));
+                String instanceIdGenerator = instance.getInstanceIdGenerator();
+                if (Constants.SNOWFLAKE_INSTANCE_ID_GENERATOR.equalsIgnoreCase(instanceIdGenerator)) {
+                    Instance oldInstance = instanceMap.get(instance.getDatumKey());
+                    if (oldInstance != null && oldInstance.isSnowflakeId()) {
+                        instance.setInstanceId(oldInstance.getInstanceId());
+                    } else {
+                        instance.setInstanceId(generateSnowflakeInstanceId(currentInstanceIds));
+                    }
+                }
```
Comment (KomachiSion) [inline naming/src/main/java/com/alibaba/nacos/naming/core/ServiceManager.java:822]: Bad implementation I think. If oldInstance exist, only use instance id of old instance, can't change instanceId.

The original implementation also has some problem

Comment (horizonzy) [reply naming/src/main/java/com/alibaba/nacos/naming/core/ServiceManager.java:822]: some situation.
1. old instance didn't use snowflake generator, new instance use snowflake generator. New instance id should be snowflake Id.
2.old instance use snowflake generator, new instance isn't use  snowflake generator. New instance id shouldn't remains old instance Id. 
3.old instance and new instance all use snowflake generator, new instance should remains old instance Id.

Comment (KomachiSion) [reply naming/src/main/java/com/alibaba/nacos/naming/core/ServiceManager.java:822]: 1. Must deregister instance and register again.

2. Same as the first

3. yes it is.

Comment (KomachiSion) [reply naming/src/main/java/com/alibaba/nacos/naming/core/ServiceManager.java:822]: Instance id is a binder like ip and port (so default id generate from ip and port), If user want to change, they **must** deregister the older and register new. Can't update it directly.

Comment (horizonzy) [reply naming/src/main/java/com/alibaba/nacos/naming/core/ServiceManager.java:822]: > Instance id is a binder like ip and port (so default id generate from ip and port), If user want to change, they **must** deregister the older and register new. Can't update it directly.

agree

**Review by KomachiSion — APPROVED** [review]

**Pull request merged**